commit.go

```diff
@@ -319,7 +333,8 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 	}
 
 	var manifestBytes []byte
-	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy), options.MaxRetries, options.RetryDelay); err != nil {
+	logrus.Debugf("Calling from Commit")
+	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy, nil, nil, nil), options.MaxRetries, options.RetryDelay); err != nil {
```

Comment:
It is possible to commit directly to a remote registry, so the commit command should probably contain the same encryption options as push.

Reply:
My impression was that commit was to create an image locally and then push to export it to a registry or files. My thought was that locally committed images would be decrypted and unpacked so that they can be run via the CLI, and so any images that are allowed to be used locally to build on top of should remain decrypted. I'm not too sure about specifics about commit.

Reply:
I agree it’s not the most common usage, but it is explicitly documented e.g. at https://github.com/containers/buildah/blob/master/docs/buildah-commit.md#example .

Reply:
Let me dig a bit deeper. There may be some weird interactions, especially with buildah run... Unless we want to special case on registry for commit?

Reply:
There is already an `options.SignBy` on this line, so I imagine the encryption options could get from the CLI to this place using a similar route. But I’m only minimally familiar with the Buildah codebase.

Reply:
Okay i think i misunderstood how the running of contianers works here.

I think it looks good, the only issue i seem to be running into an issue with `blobcache`. Works fine otherwise 

local image
```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df test
Getting image source signatures
Copying blob 5f70bf18a086 done
error committing container "contid" to "test": error copying layers and metadata for container "b606f78574df9ddfefc04d0a98f76fa2ae97a1ebaed1df5cde375576c692be8c": Error reading blob sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab: error opening file "/var/tmp/buildah112495834/sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab" to buffer layer blob: open /var/tmp/buildah112495834/sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab: no such file or directory
```

Other cases
```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df docker://localhost:5000/test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config b0924b4f50 done
Writing manifest to image destination
Storing signatures
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df oci:test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 8959f8f5d5 done
Writing manifest to image destination
Storing signatures
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df dir:test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 5ded513e65 done
Writing manifest to image destination
Storing signatures

```

Reply:
Ok i fixed the above error... but when I commit to a local image. it is trying to unpack the image afterwards. Let me see if i can commit it without unpacking.

```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 45c0e7b514 done
Writing manifest to image destination
Storing signatures
error committing container "contid" to "test": error copying layers and metadata for container "b606f78574df9ddfefc04d0a98f76fa2ae97a1ebaed1df5cde375576c692be8c": Error committing the finished image: error adding layer with blob "sha256:c41eabc97b5b0885d97663759516f07af539ba887b5d78a4e12e26b681aca7d7": Error processing tar file(exit status 1): archive/tar: invalid tar header
```

Reply:
@mtrmac  It looks like (but please correct me) that the decision to unpack is being done based on the `ImageDestination`. For this specific case, I believe it is from this function which is done during `ImageDestination.Commit` via `PutLayer`.

https://github.com/containers/image/blob/master/storage/storage_image.go#L743

It looks possible to add a check around here since the manifest is available, we can skip the unpacking of encrypted layers by checking the mediatype:

```
if strings.HasSuffix(layer.BlobInfo.Mediatype, "+encrypted") {
    continue
}
```
https://github.com/containers/image/blob/master/storage/storage_image.go#L652

The issue with this would be that not all container images that are in this particular store would be runnable. Would this affect the expected semantics of this `ImageDestination`.

Alternatively, in the meantime I can do `buildah commit` that will implement encryption for everything except local images, in which case it will throw an error on the `buildah` level

Reply:
> @mtrmac It looks like (but please correct me) that the decision to unpack is being done based on the `ImageDestination`.

By “unpack”, do you mean “decompress” or “extract an archive to individual files”?

> For this specific case, I believe it is from this function which is done during `ImageDestination.Commit` via `PutLayer`.
> 
> https://github.com/containers/image/blob/master/storage/storage_image.go#L743
> 
> It looks possible to add a check around here since the manifest is available, we can skip the unpacking of encrypted layers by checking the mediatype:
> 
> ```
> if strings.HasSuffix(layer.BlobInfo.Mediatype, "+encrypted") {
>     continue
> }
> ```

That makes no sense to me. c/storage only represents layers as filesystem trees, and images as a series of layers on top of each other. If some “encrypted” layer is not recorded as a filesystem tree, it plainly does not exist and therefore an image containing that layer can not exist in c/storage either.

It does seem that we forgot to teach c/image that some transports can only work with unencrypted layers, and refuse to copy encrypted layers without decrypting; but that should happen in the generic code, the same way transports declare `AcceptsForeignURLs` they should declare they (in?)ability to accept encrypted layers.

Either way, this does not quite help Buildah…

---

> Alternatively, in the meantime I can do `buildah commit` that will implement encryption for everything except local images, in which case it will throw an error on the `buildah` level

Yes, I think something like that should happen. It is currently, AFAIK, plainly impossible to represent an encrypted image when committing to local storage, so Buildah should not be trying. (It may well be good enough to just add the check in c/image and have Buildah report the error, but reporting a better-targeted error when the user asks for a nonsensical operation could be more helpful.)

---

(All of this being said, it’s been quite some time since I’ve last read the blob cache code, and I’m not quite sure it can work with encryption. But it seems you are testing all of that already.)